File: 78.py
# Notes
    # Can visualize outputs generated as tree where each layer is a choice of
        # including/excluding a particular value
        # e.g. first layer is include/don't include 1, yieling [1] and []
        # second layer is include/don't include 2, yielding [1,2], [1], [2], []
        # continue this process for all elements
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        subsets = []

        subset = []
        def dfs(idx = 0):
            if idx >= len(nums):
                subsets.append(subset[:])
                return

            # Include element
            subset.append(nums[idx])
            dfs(idx + 1)

            # Do not include element
            subset.pop()
            dfs(idx + 1)

        dfs()
        return subsets

File: C++/78.cpp
/*
- Unique elements - no need to sort the vector
- Steps
    - Keep a vector of current subset
    - For each element in the array, perform DFS both including/excluding the element
*/
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> cur;
        dfs(nums, 0, cur, result);
        return result;
    }
private:
    void dfs(vector<int>& nums, int start, vector<int>& cur, vector<vector<int>>& result){
        result.push_back(cur);
        for(int i = start; i < nums.size(); i++){
            cur.push_back(nums[i]);
            dfs(nums, i + 1, cur, result);
            cur.pop_back();
        }
    }
};